runtime/domain.c

```diff
@@ -45,20 +45,31 @@
 #include "caml/sync.h"
 #include "caml/weak.h"
 
-/* From a runtime perspective, domains must handle stop-the-world (STW)
-   sections, during which:
-    - they are within a section no mutator code is running
-    - all domains will execute the section in parallel
-    - barriers are provided to know all domains have reached the
-      same stage within a section
+/* The runtime can run stop-the-world (STW) sections, during which all
+   active domains run the same callback in parallel (with a barrier
+   mechanism to synchronize within the callback).
 
    Stop-the-world sections are used to handle duties such as:
     - minor GC
-    - major GC to trigger major state machine phase changes
-
-   Two invariants for STW sections:
-    - domains only execute mutator code if in the stop-the-world set
-    - domains in the stop-the-world set guarantee to service the sections
+    - major GC phase changes
+
+   Code within the STW callback can have the guarantee that no mutator
+   code runs in parallel -- precisely, the guarantee holds only for
+   code that is followed by a barrier. Furthermore, new domains being
+   spawned are blocked from running any mutator code while a STW
+   section is in progress, and terminating domains cannot stop until
+   they have participated to all STW sections currently in progress.
+
+   To provide these guarantees:
+    - Domains must register as STW participants before running any
+      mutator code.
+    - Domains registered as STW participants must be careful to
```

Comment:
Nitpick: This comment is correct, but is not needed "To provide these guarantees". The guarantee is "no mutator code runs while STW section is in progress". For this guarantee, STW participants need not service STW interrupt requests. If the STW participants don't service interrupt requests, the program won't enter STW section, and the guarantee is trivially satisfied, isn't it? 

If this understanding is correct, we may move the second item out of this list, and add it as a separate paragraph following the list. We may say "For timely handling of STW requests, <2nd item>".

Reply:
Indeed. I moved the second item to a separate paragraph, and was drawn into mentioning "poll points" and "backup thread", and I ended up moving this remark to the second comment just below which is gives an overview of the backup-thread mechanism.